File: touchup.pl
#!/usr/bin/perl

use strict;
use warnings;

use constant MAX_STRING_LENGTH => 1024;
use constant MAX_VAR_DIMS => 32;

my $input_file = $ARGV[0];
my $output_file = $ARGV[0] . ".inc";

# Attributes for functions and their arguments.
my %manual_function_attributes = (
    SDgetinfo => {
        parameter_attributes => {
            dimsizes => ["size_is(" . MAX_VAR_DIMS . ")"]
        }
    },
    SDcreate => {
        parameter_attributes => {
            dimsizes => ["size_is(rank)"]
        }
    },
);
# Functions which simply return a 1 or 0 for success vs failure.
my %returns_error_code = (
    Hclose => 1,

    Vend => 1,
    Vfinish => 1,
    Vinitialize => 1,
    Vstart => 1,

    VSdetach => 1,
    VSfdefine => 1,
    VSgetclass => 1,
    VSinquire => 1,
    VSsetclass => 1,
    VSsetfields => 1,
    VSsetinterlace => 1,
    VSsetname => 1,

    SDdiminfo => 1,
    SDend => 1,
    SDendaccess => 1,
    SDfileinfo => 1,
    SDgetinfo => 1,
);

=head2 read_file
Read in the file contents, return an array containing each of the lines.
=cut
sub read_file {
    my $file = $input_file;
    open my $fin, $file or die "Unable to open $file for reading: $!";
    my @lines = <$fin>;
    close $fin;
    return @lines;
}

=head2 clean_up
Some initial code cleanup.
This:
 * Removes comments
 * Removes preprocessor directives (though there shouldn't be any left)
 * Replaces strings of whitespace longer than 1 character with a single space
 * Puts a newline after each ;
=cut
sub clean_up {
    my $code = shift;

    # Comments
    # Taken from perlfaq6
    $code =~ s#/\*[^*]*\*+([^/*][^*]*\*+)*/|//[^\n]*|("(\\.|[^"\\])*"|'(\\.|[^'\\])*'|.[^/"'\\]*)#defined $2 ? $2 : ""#gse;
    # Preprocessor stuff
    $code =~ s/^\#.*(\\\n.*)*//mgo;
    # Shrink the whitespace
    $code =~ s/\s+/ /g;
    # Add newlines back in to separate function definitions
    $code =~ s/; /;\n/g;
    return $code;
}

=head2 make_attribute_string
Description:
Return a string containing the properly formatted text for the given array
of attributes.

Arguments:
Array of attributes to use (NOT arrayref).

Returns:
String of attributes, include surrounding square brackets.
=cut
sub make_attribute_string {
    return "[" . (join ', ', @_) . "]";
}

=head2 minimize_whitespace
Description:
Strip leading, trailing whitespace from each argument.  Also compress any
internal whitespace to a single space character.

Arguments:
Array of string to modify (they are modified in place!)

Returns:
Nothing important.
=cut
sub minimize_whitespace {
    for (@_) {
        # Strip leading
        s/^\s+//;
        # Strip trailing
        s/\s+$//;
        # Collapse multiple whitespace characters to one space
        s/\s+/ /;
    }
}

=head2 function_attributes
Description:
Given a function name and return type, determine its attributes.

Arguments:
1. Function return type
2. Function name

Returns:
=cut
sub function_attributes {
    my ($type, $name) = @_;

    my %attributes = ();

    # Remove the UPPERCASE from the OCaml names, since OCaml does not allow
    # functions to have names with leading capital letters.
    if ($name =~ /^([A-Z]+)(.*)$/) {
        $attributes{"mlname(" . lc($1) . "_$2)"} = 1;
    }

    # If returning a char *, make it a string
    if ($type =~ /char\s+\*/) {
        $attributes{"string"} = 1;
        $attributes{"length_is(" . MAX_STRING_LENGTH . ")"} = 1;
    }

    # Function-specific special cases
    if (grep { /$name/ } (keys %manual_function_attributes)) {
        print "Manual function attribute(s) available for $name\n";
        # Add each manual attribute to the attribute record hash.
        my $function_attributes = $manual_function_attributes{$name}->{function_attributes};
        for (@$function_attributes) {
            $attributes{$_} = 1;
        }
    }

    (keys %attributes) ? (keys %attributes) : ();
}

=head2 maybe_adjust_return_type
Description:
If appropriate, change the return type of the given function so that "failure"
return values raise exceptions.

Arguments:
1. Original return type of the function.
2. Function name.

Returns:
Function return type that should be used.
=cut
sub maybe_adjust_return_type {
    my ($type, $name) = @_;

    if (grep { /$name/ } (keys %returns_error_code)) {
        print "Changing return type of $name to error code\n";
        return "HDF_RESULT";
    }
    else {
        return $type;
    }
}

=head2 argument_attributes
Description:
Determine appropriate attributes for function arguments.

Arguments:
1. Function name
2. Attribute type
3. Attribute name

Returns:
Array of argument attributes
OR
undef if there are no attributes for the given argument
=cut
sub argument_attributes {
    my ($function, $type, $name) = @_;

    my %attributes = ();

    # *get* and *info* and *inquire* are (generally?) using pointer values
    # as output.
    if ($function =~ /get|info|inquire/ and $type =~ /\*/) {
        $attributes{out} = 1;
    }

    # Make (const and un-const) char * arguments usable as strings.
    if ($type =~ /(?:const )?char \*/) {
        $attributes{string} = 1;
        if ($attributes{out}) {
            $attributes{"length_is(" . MAX_STRING_LENGTH . ")"} = 1;
        }
    }

    # Function-specific special cases
    if (grep { /$function/ } (keys %manual_function_attributes)) {
        if (grep { /$name/ } (keys %{$manual_function_attributes{$function}->{parameter_attributes}})) {
            print "Manual parameter attribute(s) available for $name\n";
            # Add each manual attribute to the attribute record hash.
            my $parameter_attributes = $manual_function_attributes{$function}->{parameter_attributes}->{$name};
            for (@$parameter_attributes) {
                $attributes{$_} = 1;
            }
        }
    }

    (keys %attributes) ? [keys %attributes] : undef;
}

=head2 process_prototype
Line-by-line processing of the function prototypes.
=cut
sub process_prototype {
    my $line = shift;
    my $bigarray = shift;
    $bigarray = $bigarray ? "bigarray, " : "";

    # There is probably a better way to do this...
    my ($return_type, $function_name, $args) = $line =~ /^((?:const )?\w+ (?:\*\s*)?)(\w+)\s*\(([\w\s\*\[\],]*)\)/;
    unless ($return_type and $function_name and $args) {
        print "$line\n";
    }
    minimize_whitespace ($return_type, $function_name, $args);
    $return_type = maybe_adjust_return_type($return_type, $function_name);

    # Hold attribute information for the whole prototype
    my @function_attributes = function_attributes $return_type, $function_name;

    # Argument list, split as a hash holding name and type for each (in order)
    my @arg_list = split ',', $args;
    my @function_args = map {
        my ($type, $name) = $_ =~ /(.*)?\b(\w+)/;
        minimize_whitespace ($type, $name);
        my %arg_hash = (name => $name, type => $type);
        my $attributes = argument_attributes $function_name, $type, $name;
        if ($attributes) {
            $arg_hash{attributes} = $attributes;
        }
        \%arg_hash
    } (split ',', $args);

    my @updated_arg_list = ();
    for my $arg (@function_args) {
        my @pieces = ($arg->{type}, $arg->{name});
        if (defined $arg->{attributes}) {
            @pieces = ((make_attribute_string @{$arg->{attributes}}), @pieces);
        }
        push @updated_arg_list, (join ' ', @pieces);
    }

    # Join it all back together, (hopefully) ready for camlidl processing.
    return
        (@function_attributes ? (make_attribute_string @function_attributes) . " " : "")
        . "$return_type $function_name("
        . (join ', ', @updated_arg_list) . ");";
}

my @input_lines = read_file ();
my $blob = join "", @input_lines;
my @scrubbed_lines = split "\n", clean_up $blob;

sub handle_file {
    my $bigarray = shift;
    my $output_filename = shift;

    my $OUTFILE;
    open $OUTFILE, ">$output_filename" or die "Error with output file: $!";

    my $is_a_typedef_struct = 0;
    my $is_a_typedef_enum = 0;

    for (@scrubbed_lines) {
        my $line = $_;
        SWITCH: {
            # Don't do processing inside of typedef (struct) definitions.
            # XXX - This really is a horrible hack, but it does the job.
            if (/^\s*typedef\s+struct[\s\w]*{/) { $is_a_typedef_struct = 1; last SWITCH; }
            if (/^\s*typedef.*?;/) { last SWITCH; }
            if (/}/ and $is_a_typedef_struct) { $is_a_typedef_struct = 0; last SWITCH; }
            if ($is_a_typedef_struct) { last SWITCH; }
            # If we make it to here, then we're not in a typedef, so we should process
            # the line.
            $line = process_prototype $_, $bigarray;
        }

        print $OUTFILE "$line\n";
    }

    close $OUTFILE;
}

handle_file 0, $output_file;

File: touchup.ml
#use "topfind";;
#require "unix";;
#require "pcre";;

type attribute_spec = {
  function_name: string;
  function_attrs: string list option;
  parameter_attrs: (string * string list) list option;
}

let max_var_dims = 32

let manual_function_attributes =
  [
    {
      function_name = "SDgetinfo";
      function_attrs = None;
      parameter_attrs =
        Some ["dimsizes", ["size_is(" ^ string_of_int max_var_dims ^ ")"]];
    };
    {
      function_name = "SDcreate";
      function_attrs = None;
      parameter_attrs = Some ["dimsizes", ["size_is(rank)"]];
    };
    {
      function_name = "SDsetcompress";
      function_attrs = None;
      parameter_attrs = Some ["c_info", ["in"; "ref"]];
    };
    {
      function_name = "VSfindex";
      function_attrs = None;
      parameter_attrs = Some ["field_index", ["out"]];
    }
  ]

(* Functions which simply return a 1 or 0 for success vs failure. *)
let returns_error_code =
  [
    "Hclose";
    "Vend";
    "Vfinish";
    "Vinitialize";
    "Vstart";

    "VSattrinfo";
    "VSdetach";
    "VSfdefine";
    "VSfindex";
    "VSgetclass";
    "VSinquire";
    "VSsetclass";
    "VSsetfields";
    "VSsetinterlace";
    "VSsetname";

    "SDattrinfo";
    "SDdiminfo";
    "SDend";
    "SDendaccess";
    "SDfileinfo";
    "SDgetinfo";
    "SDgetcompress";
    "SDsetcompress";
    "SDgetdatastrs";
    "SDsetdatastrs";
  ]
let hdf_error_return = "HDF_RESULT"


(* Length to allocate for output strings. *)
let max_string_length = "1024"

(* Functions to read in everything on STDOUT from a given command. *)
(* Many thanks to Richard M. Jones for the following two functions! *)
let rec input_all_lines chan =
  try
    let line = input_line chan in
    line :: input_all_lines chan
  with
      End_of_file -> []

let pget cmd =
  let chan = Unix.open_process_in cmd in
  let lines = input_all_lines chan in
  let stat = Unix.close_process_in chan in
  (match stat with
       Unix.WEXITED 0 -> ()
     | Unix.WEXITED i ->
         failwith ("command failed with code " ^ string_of_int i)
     | Unix.WSIGNALED i ->
         failwith ("command killed by signal " ^ string_of_int i)
     | Unix.WSTOPPED i ->
         failwith ("command stopped by signal " ^ string_of_int i));
  lines

let read_file filename =
  let preprocessed_text = pget ("cpp " ^ filename) in
  let l = List.map (fun l -> l ^ "\n") preprocessed_text in
  l

let (|>) x f = f x
let id x = x

let cleanup_lines l =
  (* Strip out #-started preprocessor lines, as well as lines with only
     whitespace. *)
  let blob =
    let filtered = 
      List.filter (
        fun line ->
          if Pcre.pmatch ~pat:"^#|^\\s+$" line then
            false
          else
            true
      ) l
    in
    List.fold_left (^) "" filtered
  in
  blob
    (* Compress lengths of whitespace down to a single character *)
    |> Pcre.replace ~pat:"\\s+" ~templ:" "
    (* Put newlines back in after each ; *)
    |> Pcre.replace ~pat:"; " ~templ:";\n"

let make_attribute_string attributes =
  match attributes with
      [] -> ""
    | a ->
        "[" ^ String.concat ", " a ^"]"

let minimize_whitespace s =
  s
    |> Pcre.replace ~pat:"^\\s+" ~templ:""
    |> Pcre.replace ~pat:"\\s$" ~templ:""
    |> Pcre.replace ~pat:"\\s+" ~templ:" "

let function_attributes return_type name =
  let check_re re =
    if Pcre.pmatch ~pat:re name then
      Pcre.extract ~pat:re ~full_match:false name
    else
      [||]
  in

  let name_checks =
    [
      (* OCaml values can not begin with a capital letter.  Translate a name
         like FOObar to foo_bar for OCaml. *)
      "^([A-Z]+)(.*)$",
      (
        fun a -> ["mlname(" ^ (
          match Array.length a with
              1 -> String.lowercase a.(0)
            | 2 ->
                String.lowercase a.(0) ^ "_" ^ a.(1)
            | _ -> raise (Failure "Bad result in function caps check")
        ) ^ ")"]
      );
      (* Plplot names many of their functions c_* to avoid clashes with certain
         language bindings.  There's no need to carry this over to OCaml.
         This turns c_foo in to foo. *)
      "^c_(\\w+)$", (fun a -> ["mlname(" ^ a.(0) ^ ")"]);
    ]
  in
  let type_checks =
    [
      (* Treat strings properly *)
      "char\\s*\\*",
      ["string"; "length_is(" ^ max_string_length ^ ")"]
    ]
  in

  (* Attributes based on the function name *)
  let name_attrs =
    List.map
      (fun (re,attrf) -> let a = check_re re in if Array.length a > 0 then attrf a else [])
      name_checks
    |> List.flatten
  in
  (* Attributes based on the function type *)
  let type_attrs =
    List.map
      (fun (re,attrs) -> if Pcre.pmatch ~pat:re return_type then attrs else [])
      type_checks
    |> List.flatten
  in
  (* Any other attributes, specified manually *)
  let manual_attrs =
    try
      let fa =
        List.find (fun fa -> fa.function_name = name) manual_function_attributes
      in
      match fa.function_attrs with
          Some a -> a
        | None -> []
    with
        Not_found -> []
  in
  name_attrs @ type_attrs @ manual_attrs

let parameter_attributes function_name types names =
  let pmatch re str = Pcre.pmatch ~pat:re str in
  let non_get_functions = ["c_plgriddata"; "c_plgra"] in

  (* If all of the pieces are true, then the attribute(s) is(are) appropriate
     for this parameter. *)
  let checks p_type p_name =
    [
      (* Order goes:
         function_name check
         type check
         attribute name check
         misc. check (anything, as long as it's a bool)
         attributes, if all of the above are true
      *)
      (* "get" functions *)
      pmatch "get|info|inquire" function_name,
      pmatch "\\*" p_type,
      true,
      not (List.mem function_name non_get_functions),
      ["out"] @
        if pmatch "char" p_type then ["length_is(" ^ max_string_length ^ ")"]
        else [];
      (* Strings *)
      true,
      pmatch "(?:const )?char\\s*\\*$" p_type,
      true,
      true,
      ["string"];
    ]
  in

  let attr_hash = Hashtbl.create 10 in

  let perform_check param_type param_name =
    (* Any other attributes, specified manually *)
    let manual_attrs =
      try
        let fa =
          List.find (fun fa -> fa.function_name = function_name) manual_function_attributes
        in
        match fa.parameter_attrs with
            Some a -> 
              List.assoc param_name a
          | None -> []
      with
          Not_found -> []
    in
    Hashtbl.add attr_hash param_name manual_attrs;
    (* Check for attributes, filter the ones we don't want, then add the rest
       to the attribute hash. *)
    checks param_type param_name
    |> List.filter (
        fun (function_check, type_check, name_check, other_check, _) ->
          List.for_all id [function_check; type_check; name_check; other_check]
       )
    |> List.iter (fun (_,_,_,_,attrs) -> Hashtbl.add attr_hash param_name attrs)
  in

  List.iter2 perform_check types names;
  attr_hash

let build_attribute_list l =
  List.map (
    fun (attrs, t, n) ->
      String.concat " " [make_attribute_string attrs; t; n]
  ) l

let process_prototype line =
  let pieces =
    line
    |> Pcre.extract ~pat:"^((?:const )?\\w+ (?:\\*\\s*)?)(\\w+)\\s*\\(([\\w\\s\\*\\[\\],]*)\\)" ~full_match:false
    |> Array.map minimize_whitespace
  in
  (* Get the return type, name and arg list separately *)
  let return_type =
    if List.mem pieces.(1) returns_error_code then
      hdf_error_return
    else
      pieces.(0)
  in
  let function_name = pieces.(1) in
  let params =
    pieces.(2)
    |> Pcre.split ~pat:","
    |> List.map minimize_whitespace
  in
  let param_types, param_names =
    params
    |> List.map (
         fun param ->
           let p = Pcre.extract ~pat:"(.*)?\\b(\\w+)" ~full_match:false param in
           minimize_whitespace p.(0), minimize_whitespace p.(1)
       )
    |> List.split
  in
  let f_attrs = function_attributes return_type function_name in
  let p_attrs = parameter_attributes function_name param_types param_names in
  let params_with_attrs =
    List.map2
      (fun t n -> Hashtbl.find_all p_attrs n |> List.flatten, t, n)
      param_types param_names
  in
  String.concat " " (
    [
      make_attribute_string f_attrs;
      return_type;
      function_name; "(";
    ]
    @ [String.concat ", " (build_attribute_list params_with_attrs)]
    @ [");"]
  )

let write_file filename lines =
  let fout = open_out filename in
  List.iter (output_string fout) lines;
  close_out fout;
  ()

let process_file filename =
  let lines = read_file filename in
  let lines' = cleanup_lines lines |> Pcre.split ~pat:"\n" in
  lines'
    |> List.map (
         fun l -> try process_prototype l with Not_found -> l
       )
    |> List.map minimize_whitespace
    |> List.map (fun l -> l ^ "\n")
    |> write_file (filename ^ ".inc")

let () =
  if !Sys.interactive then
    ()
  else (
    process_file "hdf_h";
    process_file "mfhdf_h";
    ()
  )
